### PR TITLE
Namespace controller should always get latest state prior to deletion

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Multiple controllers may edit a `Namespace` during termination that result in resource version changes if there are multiple finalizers.  In order to optimize evacuation from potential resource conflicts, we first fetch the latest state of the `Namespace` if its in the process of deletion to try and have an updated state because of the latency around watch notification.

/cc @smarterclayton

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
marking wip while I fix a test case.

**Comment by smarterclayton** [comment]
We're using framework.Controller as an Informer but it doesn't have delta compression - so we get every update that happens while the process of namespace deletion is going on.  @lavalamp should we be able to create a NewInformer for the namespace controller that has delta compression?

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -305,6 +317,10 @@ func syncNamespace(kubeClient client.Interface, versions *unversioned.APIVersion
 	// we have removed content, so mark it finalized by us
 	result, err := retryOnConflictError(kubeClient, namespace, finalizeNamespaceFunc)
 	if err != nil {
+		// it was removed already, so life is good
```
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: Could we move this into finalizeNamespaceFunc to return nil?

**Comment by derekwaynecarr** [comment]
@smarterclayton - comments addressed.

event with compression (which we should do later), i think we will still want this to avoid resource conflicts due to latency in watch.  event compression will help avoiding additional calls in UpdateFunc after the actual delete happens though.  i think that can be a follow-on.

**Comment by lavalamp** [comment]
You can provide a delta compresser function to just toss things you don't
care about.

**Comment by smarterclayton** [comment]
We'd have to change the signature of NewInformer, I just wanted to be sure there wasn't a reason you had it that way.

**Comment by derekwaynecarr** [comment]
This PR should be merged as is for-now.

This PR will improve how the function works in core Kubernetes independent.

Right now, a delete of a namespace in Kubernetes results in the following:

```
UpdateFunc (namespace to be deleted)
   == finalize the namespace
   == delete all content
   == delete the namespace
UpdateFunc (namespace was finalized)
   == report error in log (namespace was already deleted)
```

with this change:

```
UpdateFunc (namespace to be deleted)
   == finalize the namespace
   == delete all content
   == delete the namespace
UpdateFunc (namespace was finalized)
   == return (namespace was deleted)
```

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by eparis** [comment]
@wojtek-t is the merge bot acting up? if this didn't have a clean shippable or jenkins unit, it shouldn't have gone in, right?

**Comment by wojtek-t** [comment]
@eparis IIUC merge bot is acting fine. What happened here is the "test this please" command is currently invoking both e2e and unit tests. The latter one failed, but I managed to trigger it again before merge-bot observed it. The second run finished correctly and it was merged.

**Pull request merged**